File: app/src/main/java/com/sliide/news/core/viewmodel/NewsListViewModel.java
package com.sliide.news.core.viewmodel;

import android.view.View;
import android.widget.LinearLayout;

import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.SimpleItemAnimator;

import com.sliide.news.core.datamodel.INewsListDataModel;
import com.sliide.news.core.datamodel.NewsListDataModelImpl;
import com.sliide.news.network.model.NewsItem;
import com.sliide.news.network.model.NewsResponseModel;

import java.util.ArrayList;

import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

public class NewsListViewModel {

    private RecyclerView recyclerView;
    private NewsRecyclerViewAdapter newsRecyclerViewAdapter;
    private ArrayList<NewsItem> result = new ArrayList<>();
    private ViewPagerAdapter pagerAdapter;


    protected NewsRecyclerViewAdapter getAdapter() {
        if (pagerAdapter == null) {
            pagerAdapter = new ViewPagerAdapter(result);
        }
        if (newsRecyclerViewAdapter == null) {
            newsRecyclerViewAdapter = new NewsRecyclerViewAdapter(result, pagerAdapter);
        }
        return newsRecyclerViewAdapter;
    }


    protected RecyclerView.LayoutManager setLayoutManager() {
        if (getRecyclerView() != null) {
            getRecyclerView().addItemDecoration(new DividerItemDecoration(getRecyclerView().getContext(), LinearLayout.VERTICAL));
        }
        return new LinearLayoutManager(getRecyclerView().getContext());
    }

    private INewsListDataModel newsListDataModel;

    public NewsListViewModel() {
        newsListDataModel = new NewsListDataModelImpl();
        addListToView(null);
    }

    public void addListToView(View view) {

        newsListDataModel.getNewsList()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<NewsResponseModel>() {
                    @Override
                    public void onSubscribe(Disposable d) {

                    }

                    @Override
                    public void onNext(NewsResponseModel newsResponseModel) {
                        result.clear();
                        pagerAdapter = new ViewPagerAdapter(result);
                        result.addAll(newsResponseModel.content);
                        if (newsRecyclerViewAdapter != null) {
                            newsRecyclerViewAdapter.notifyDataSetChanged();
                            pagerAdapter.notifyDataSetChanged();
                        }
                    }

                    @Override
                    public void onError(Throwable e) {

                    }

                    @Override
                    public void onComplete() {

                    }
                });

    }

    public final void setupRecyclerView(RecyclerView recyclerView) {
        this.recyclerView = recyclerView;
        this.recyclerView.setHasFixedSize(true);
        this.recyclerView.setAdapter(getAdapter());
        this.recyclerView.setLayoutManager(setLayoutManager());
        this.recyclerView.setOverScrollMode(View.OVER_SCROLL_NEVER);
        if (this.recyclerView.getItemAnimator() != null && this.recyclerView.getItemAnimator() instanceof SimpleItemAnimator) {
            ((SimpleItemAnimator) this.recyclerView.getItemAnimator()).setSupportsChangeAnimations(false);
        }
    }

    public RecyclerView getRecyclerView() {
        return recyclerView;
    }

}

File: app/src/main/java/com/sliide/news/core/viewmodel/NewsListViewModel.kt
package com.sliide.news.core.viewmodel

import android.view.View
import android.widget.LinearLayout
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.SimpleItemAnimator
import com.sliide.news.core.datamodel.INewsListDataModel
import com.sliide.news.core.datamodel.NewsListDataModelImpl
import com.sliide.news.network.model.NewsItem
import com.sliide.news.network.model.NewsResponseModel
import io.reactivex.Observer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import java.util.ArrayList

class NewsListViewModel {
    private var recyclerView: RecyclerView? = null
    private var newsRecyclerViewAdapter: NewsRecyclerViewAdapter? = null
    private val result: ArrayList<NewsItem> = ArrayList()
    private var pagerAdapter: ViewPagerAdapter? = null

    protected fun getAdapter(): NewsRecyclerViewAdapter {
        if (pagerAdapter == null) {
            pagerAdapter = ViewPagerAdapter(result)
        }
        if (newsRecyclerViewAdapter == null) {
            newsRecyclerViewAdapter = NewsRecyclerViewAdapter(result, pagerAdapter!!)
        }
        return newsRecyclerViewAdapter!!
    }

    protected fun setLayoutManager(): RecyclerView.LayoutManager {
        getRecyclerView()?.addItemDecoration(
            DividerItemDecoration(
                getRecyclerView()!!.context,
                LinearLayout.VERTICAL
            )
        )
        return LinearLayoutManager(getRecyclerView()!!.context)
    }

    private val newsListDataModel: INewsListDataModel

    init {
        newsListDataModel = NewsListDataModelImpl()
        addListToView(null)
    }

    fun addListToView(view: View?) {
        newsListDataModel.getNewsList()
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(object : Observer<NewsResponseModel> {
                override fun onSubscribe(d: Disposable) {
                    // Not implemented
                }

                override fun onNext(newsResponseModel: NewsResponseModel) {
                    result.clear()
                    pagerAdapter = ViewPagerAdapter(result)
                    newsResponseModel.content?.let { result.addAll(it) }
                    newsRecyclerViewAdapter?.let {
                        it.notifyDataSetChanged()
                        pagerAdapter?.notifyDataSetChanged()
                    }
                }

                override fun onError(e: Throwable) {
                    // Not implemented
                }

                override fun onComplete() {
                    // Not implemented
                }
            })
    }

    fun setupRecyclerView(recyclerView: RecyclerView) {
        this.recyclerView = recyclerView
        this.recyclerView?.setHasFixedSize(true)
        this.recyclerView?.adapter = getAdapter()
        this.recyclerView?.layoutManager = setLayoutManager()
        this.recyclerView?.overScrollMode = View.OVER_SCROLL_NEVER
        if (this.recyclerView?.itemAnimator is SimpleItemAnimator) {
            (this.recyclerView?.itemAnimator as SimpleItemAnimator).supportsChangeAnimations = false
        }
    }

    fun getRecyclerView(): RecyclerView? {
        return recyclerView
    }
}